### DISCUSSION TITLE
Backoffice support for front end accessibility

### DISCUSSION THREAD

**Discussion by RachBreeze** [umbraco/Umbraco-CMS#10468] [Features and ideas]
Hello

The Umbraco Backoffice accessibility initiative has done and continues to do a great job on improving the accessibility of the Backoffice.  

There are a number of features which can be added to Umbraco which can help create accessible content by default and which can be added to the solution to aid content authors when creating accessible implementations.

### Language of parts
When an element changes from the natural language of the page to another language the html should be marked up correctly. Eg when the language of the page changes from English to French:

`<html lang="en-GB>`
....
```
<p>Hello</p>
<p lange="fr">Bonjour</p>

```
A demo of the impact of this change is found here https://vimeo.com/210258413. 

The requirement for supporting language of parts in the WCAG criteria is found here https://www.w3.org/TR/WCAG21/#language-of-parts.
A pull request for programmatically supporting language of parts can be found here https://github.com/umbraco/Umbraco-CMS/pull/9576
In addition the TinyMCE should also be updated to support language of parts, so that the user can hover over the content and change the language to the correct language. An example implementation can be found here https://github.com/joomla/joomla-cms/pull/30939 

### Alt Text Documentation 
Lots of people have trouble writing good alt text. We can improve this by providing a help page within Umbraco which documents some best practices. As part of the We4Authors Cluster https://accessibilitycluster.com one recommendation is to provide clear help text that is linked to from alt text fields. https://www.funka.com/en/projekt/cluster-feature-testing/feature-3-of-10-documentation/ 

### Alt Text Field
Alt text is context sensitive and should be added by the content author when an image is picked for display in a page.
Alt text is mandatory apart from when the image is decorative (however even when the image is decorative, having alt text does aid users on slow connections). The issue is there is no way to differentiate between an editor saying the image is decorative and missing to add the alt text. The cluster has suggested adding a check box this as a method for differentiating 
![image](https://user-images.githubusercontent.com/10153922/121913488-58acb980-cd29-11eb-883a-1d7c73f317a7.png)

### Testing accessibility in the editor
The cluster has suggested testing the accessibility of the page as the author enters content, because of how pages tend to be built in Umbraco this is not feasible. Whilst the architecture of the tool does not allow in page editor testing of the whole of the content on a page upgrading to TinyMCE 5 may provide some organisations the chance to pay for the add on for accessibility checking at an RTE level https://www.tiny.cloud/docs/enterprise/accessibility/ 

### Testing accessibility in Preview Mode
Based on this PR from joomla investigate enable a plug in https://github.com/joomla/joomla-cms/pull/31090 look to perform a full accessibility check on the page in preview mode.

I appreciate there's a few changes here to discuss, it would be great to get feedback on them.

**Comment by nul800sebastiaan** [comment]
Thanks @RachBreeze! 

> In addition the TinyMCE should also be updated to support language of parts

What does this mean? Updated to v5, or.. otherwise?

In any case, we are thinking about how we can introduce Tiny v5 at the moment, I'm not sure when this can happen. But when it does, most all of these are no-brainers to implement! 👍 

Not sure how the testing could be done, but open to ideas. We have talked before about making the preview page more extensible, which would make so much sense here!

**Reply by RachBreeze** [reply]
Hi @nul800sebastiaan 
Sorry for the confusion this means adding support for language of parts, so that if a the language of the page is changing part way  through the page the screen reader can pronounce the text correctly in the native language a video can be found here https://vimeo.com/210258413 an example may be where the user changes the welcome text from one language to another as in this demo from Washington University http://www.washington.edu/accesscomputing/AU/after.html here the page content is marked up as ``<html lang="en">``

But further down the page the content is in Spanish so the paragraph is marked :
``<p lang="es">``
We should allow editors in the tiny mce to highlight text to change the language an example implementation is found here An example implementation can be found here joomla/joomla-cms#30939

I have also got a related PR for accessibility around changing the language when the page falls back from say a Spanish page to an English piece of content, this PR is here https://github.com/umbraco/Umbraco-CMS/pull/9576

**Comment by marcemarc** [comment]
Having that 'this image is decorative' checkbox as part of the media picker would be a massive step into making people 'think' at the point of insertion and consider alt text as a thing...

It immediately explains the legitimacy of 'some images' not requiring alt text (and as a consequence flags the need for images that do)- and devs can use it to wire up role="presentation" etc

WIth this embedded in the CMS it would really support the notion of "oh this is something we need to be doing" in people's minds, the tool is pointing us in the right direction, it's a thing everybody else does, rather than, Umbraco doesn't do this, so probably nobody really needs to be doing this.

Too often I see sites 'pass' accessibility audits by 'just' using the name of the media item as the alt text on every image... and it's awful to screen read!

**Reply by RachBreeze** [reply]
It would also be "nice" if there was a way to hook ai into the media picker so that the image (possibly with content from the page) and the page language could be passed to ai to help find suitable alt text for the image.

The use of APIs calls here would mean developers could look to use Microsoft Cognitive Services / Amazon Rekognition/ Vision AI and others where applicable to aid content editors with suggested text

**Comment by DanDiplo** [comment]
For the language, you can currently just do this in Razor views/partials and it renders the current language:

`<html lang="@Culture">`

**Reply by RachBreeze** [reply]
Hi @DanDiplo thank you I was referring to the language of parts in a page, not the page content itself I've given a fuller outline in this comment https://github.com/umbraco/Umbraco-CMS/discussions/10468#discussioncomment-916221

**Reply by deanleigh** [reply]
We have done this in a live site @RachBreeze - I have written a longer explanation as a new comment.

**Reply by RachBreeze** [reply]
Hi @deanleigh  your implementation on https://www.triland.com/ was exactly what we were looking for thank you

**Comment by deanleigh** [comment]
Hi @RachBreeze Re: Language of parts
we have implemented the lang attribute in the following site:
[https://www.triland.com/](https://www.triland.com/)

We have enabled this in TinyMCE in such as  way that we can create whole blocks marked up with lang but also sentences, words or even single characters. This means we can have a mixture of languages in a single sentence.

I would very much like to be involved in moving TinyMCE's capabilities forward for Umbraco as we have using customised instances of it in our own CMS for 10+ years.

I have spoken with @sebastian-claesson at one of the hackathons in the past and he had some good ideas for quick wins to create instances within the existing framework.

I also asked @warrenbuckley about this as I saw something about 'not' upgrading to TinyMCE 5 in the move to version 9.

Regardless I would like to start a proposal for the move to version 5 and to be able to save a separate TinyMCE config for each instance which will give us access to lots of customisation as well as all the plugins.

Should I make a new proposal as I think this goes way beyond accessibility and would be a considerable mini-project in it's own right?

**Reply by RachBreeze** [reply]
Hi @deanleigh 

Your implementation on the front end of triland is exactly what we were thinking of thank you. It would be great if all Umbraco websites could do something similar OOB.

When I discussed updating TinyMCE with @nul800sebastiaan for this functionality he did say this should be after TinyMCE 5 but it does sound like that is a long way off.

As part of the a11y work I spoke to TinyMCE developers about implementing https://www.tiny.cloud/docs/enterprise/accessibility/ in TinyMCE 4 but they indicated, for support reasons, that this would be better done in TinyMCE 5

**Reply by deanleigh** [reply]
>> Your implementation on the front end of triland is exactly what we were thinking of thank you. It would be great if all Umbraco websites could do something similar OOB.

I think adding lang support could be added easily is that something to move forward with here or a new proposal? Sorry I don't really know the protocaol.

**Reply by deanleigh** [reply]
>> TinyMCE 4 but they indicated, for support reasons, that this would be better done in TinyMCE 5

I think if we were to try to review how RTE's are saved it should be in conjunction with upgrading TinyMCE anyway.
I wonder if we could get an RTE working group going to meet and write a clear requirements / proposal

**Reply by RachBreeze** [reply]
There was an RFC made for upgrading https://github.com/umbraco/rfcs/blob/main/cms/0016-TinyMCE-RTE-Improvements.md (I think this is closed now)

**Reply by RachBreeze** [reply]
Hi @deanleigh  and @warrenbuckley 
In the wrap up meeting for the https://accessibilitycluster.com/ project we've been involved in Tiny advised they have moved the language of parts functionality into their core project and are looking to move more functionality into there to help create accessible content, as Dean says it would be good to get the upgrade to TinyMCE 5 back up and running if possible.

**Reply by deanleigh** [reply]
Thank you for this @RachBreeze - I found this example fantastic as well:
[https://accessibilitycluster.com/main-results/change-language](https://accessibilitycluster.com/main-results/change-language)
